akvo/rsr/usecases/run_iati_validation_jobs.py

```diff
@@ -0,0 +1,238 @@
+        return True
+
+    def _apply_rate(self, rate: RequestRate, regulate: bool = False) -> bool:
+        if rate.count == 0 or rate.period.total_seconds() == 0:
+            return True
+        finished_at = now() - rate.period
+        count, latest = self._get_finished_jobs_summary(finished_at)
+        if count >= rate.count:
+            return False
+        if not regulate:
+            return True
+        # Spread the request evenly in the given time period
+        spacing = timedelta(seconds=rate.period.total_seconds() / rate.count)
+        if latest and latest >= now() - spacing:
+            return False
+        return True
+
+    def _get_finished_jobs_summary(self, finished_at: datetime) -> Tuple[int, Optional[datetime]]:
+        activity_jobs = IatiActivityValidationJob.objects.filter(finished_at__gt=finished_at)
+        organisation_jobs = IatiOrganisationValidationJob.objects.filter(finished_at__gt=finished_at)
+        job_count = activity_jobs.count() + organisation_jobs.count()
+        datetimes = [j.finished_at for j in [
+            activity_jobs.order_by('-finished_at').first(),
+            organisation_jobs.order_by('-finished_at').first()
+        ] if j]
+        latest = max(datetimes) if datetimes else None
+        return (job_count, latest)
+
+
+rate_limiter = IATIValidatorRateLimiter(
+    short_term=RequestRate(
+        count=getattr(settings, 'IATI_VALIDATOR_RATE_LIMIT_COUNT', 0),
+        period=getattr(settings, 'IATI_VALIDATOR_RATE_LIMIT_PERIOD', timedelta(seconds=0))
+    ),
+    long_term=RequestRate(
+        count=getattr(settings, 'IATI_VALIDATOR_QUOTA_COUNT', 0),
+        period=getattr(settings, 'IATI_VALIDATOR_QUOTA_PERIOD', timedelta(seconds=0))
+    )
+)
+
+
+def run_iati_activity_validation_job(scheduled_at: datetime = now()):
+    if not rate_limiter.is_allowed():
+        return
+    pending_jobs = get_pending_activity_jobs(scheduled_at)
+    if not pending_jobs.exists():
+        return
+    job = pending_jobs.first()
+    project = job.project
+    check_result = run_internal_project_validator(project)
+    job.mark_started()
+    try:
+        validator_result = validator.validate(get_iati_activity_xml_doc(project))
+        job.mark_finished()
+        process_activity_validation_results(project, validator_result, check_result)
+    except IATIValidatorTimeoutException:
+        pass  # retry on next round
+    except IATIValidatorException:
+        logger.exception(f'Failed to run IATI validator for project: {project.title}')
+
+
+def get_pending_activity_jobs(scheduled_at: datetime) -> QuerySet:
+    # Add 10 seconds to avoid race conditions for the failed jobs
+    timeout_at = now() - timedelta(seconds=VALIDATOR_TIMEOUT + 10)
+    return IatiActivityValidationJob.objects\
+        .select_related('project')\
+        .filter(scheduled_at__lte=scheduled_at, finished_at__isnull=True)\
+        .exclude(Q(attempts__gte=VALIDATOR_MAX_ATTEMPTS) | Q(started_at__gte=timeout_at))
```

Comment:
This is nearly identical to `get_pending_organisation_jobs`. IMO it should be merged to something like

```py
def get_pending_activity_jobs(scheduled_at: datetime) -> QuerySet:
    return _get_pending_jobs(IatiActivityValidationJob.objects.select_related('project'))

def _get_pending_jobs(queryset):
    # Add 10 seconds to avoid race conditions for the failed jobs
    timeout_at = now() - timedelta(seconds=VALIDATOR_TIMEOUT + 10)
    return queryset\
        .filter(scheduled_at__lte=scheduled_at, finished_at__isnull=True)\
        .exclude(Q(attempts__gte=VALIDATOR_MAX_ATTEMPTS) | Q(started_at__gte=timeout_at))
```

And the "magic" 10 should be moved to a constant. Unless I'm mistaken, I see it in many tests.

Reply:
Fixed